File: CoreTransformer/OutputStockCollection.vb
'************************************************************************************
' STSimStockFlow: A .NET library for simulating stocks and flows
'
' Copyright © 2009-2015 ApexRMS.
'
'************************************************************************************

Imports SyncroSim.Common
Imports System.Collections.ObjectModel

Friend Class OutputStockCollection
    Inherits KeyedCollection(Of FourIntegerLookupKey, OutputStock)

    Public Sub New()
        MyBase.New(New FourIntegerLookupKeyEqualityComparer)
    End Sub

    Protected Overrides Function GetKeyForItem(ByVal item As OutputStock) As FourIntegerLookupKey

        Return New FourIntegerLookupKey(
            item.StratumId,
            GetNullableKey(item.SecondaryStratumId),
            item.StateClassId,
            item.StockTypeId)

    End Function

End Class

File: src/Runtime/OutputStockCollection.cs
// stsim-stockflow: SyncroSim Add-On Package (to stsim) for integrating stocks and flows into state-and-transition simulation models in ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using SyncroSim.Common;
using System.Collections.ObjectModel;

namespace SyncroSim.STSimStockFlow
{
	internal class OutputStockCollection : KeyedCollection<FiveIntegerLookupKey, OutputStock>
	{
		public OutputStockCollection() : base(new FiveIntegerLookupKeyEqualityComparer())
		{
		}
		protected override FiveIntegerLookupKey GetKeyForItem(OutputStock item)
		{
			return new FiveIntegerLookupKey(
                item.StratumId,
                LookupKeyUtilities.GetOutputCollectionKey(item.SecondaryStratumId),
                LookupKeyUtilities.GetOutputCollectionKey(item.TertiaryStratumId), 
                item.StateClassId, 
                item.StockTypeId);
		}
	}
}